File: md-ml-py/matmul_party_0.py
"""Matrix multiplication experiment - party 0."""

import sys
import time
import numpy as np
from md_ml.share import Spdz2kShare64
from md_ml.protocols import PartyWithFakeOffline, Circuit

FAKE_OFFLINE_DIR = "../fake-offline-data"
JOB_NAME = "MatMul-py"
DIM = 4096
PORT = 7070


def main():
    share = Spdz2kShare64

    rng = np.random.default_rng()
    mat_a = rng.integers(0, 2**64, size=DIM * DIM, dtype=np.uint64)
    mat_b = rng.integers(0, 2**64, size=DIM * DIM, dtype=np.uint64)

    # Compute expected result locally (wrapping u64 arithmetic via numpy)
    print(f"Computing {DIM}x{DIM} reference matmul...", flush=True)
    t = time.time()
    expected = (mat_a.reshape(DIM, DIM).astype(np.uint64) @ mat_b.reshape(DIM, DIM).astype(np.uint64)).ravel()
    print(f"Reference matmul done in {(time.time()-t)*1000:.0f} ms", flush=True)

    party = PartyWithFakeOffline(share, 0, 2, PORT, JOB_NAME, FAKE_OFFLINE_DIR)
    circuit = Circuit(share)

    a = circuit.input(0, DIM, DIM)
    b = circuit.input(0, DIM, DIM)
    c = circuit.multiply(a, b)
    d = circuit.output(c)
    circuit.add_endpoint(d)

    a.set_input(mat_a, 0, share)
    b.set_input(mat_b, 0, share)

    circuit.read_offline_from_file(party)
    circuit.run_online_with_benchmark(party)
    circuit.print_stats(party)

    # Correctness check
    result = d.get_clear(share)
    if isinstance(result, np.ndarray):
        errors = int(np.sum(result != expected))
        if errors > 0 and errors < 10:
            mismatches = np.where(result != expected)[0]
            for idx in mismatches[:10]:
                row, col = divmod(int(idx), DIM)
                print(f"MISMATCH at ({row}, {col}): got {result[idx]}, expected {expected[idx]}")
    else:
        errors = sum(1 for g, e in zip(result, expected) if g != e)

    if errors == 0:
        print(f"Correctness check PASSED: all {len(expected)} elements match")
    else:
        print(f"Correctness check FAILED: {errors}/{len(expected)} elements wrong")


if __name__ == "__main__":
    main()

File: md-ml-rs/src/bin/matmul_party_0.rs
// Matrix multiplication experiment - party 0 (Rust rewrite)

use md_ml::share::Spdz2kShare64;
use md_ml::protocols::{PartyWithFakeOffline, Circuit};
use rand::Rng;

const FAKE_OFFLINE_DIR: &str = "../fake-offline-data";
const JOB_NAME: &str = "MatMul";
const DIM: usize = 4096;

fn main() {
    type S = Spdz2kShare64;
    type ClearType = u64;

    let mut rng = rand::thread_rng();
    let mat_a: Vec<ClearType> = (0..DIM * DIM).map(|_| rng.gen::<ClearType>()).collect();
    let mat_b: Vec<ClearType> = (0..DIM * DIM).map(|_| rng.gen::<ClearType>()).collect();

    // Compute expected result locally (wrapping u64 arithmetic)
    let expected = {
        let mut out = vec![0u64; DIM * DIM];
        for i in 0..DIM {
            for k in 0..DIM {
                let a = mat_a[i * DIM + k];
                for j in 0..DIM {
                    out[i * DIM + j] = out[i * DIM + j].wrapping_add(a.wrapping_mul(mat_b[k * DIM + j]));
                }
            }
        }
        out
    };

    let mut party = PartyWithFakeOffline::<S>::new(0, 2, 6060, JOB_NAME, FAKE_OFFLINE_DIR);
    let mut circuit = Circuit::<S>::new();

    let a = circuit.input(0, DIM, DIM);
    let b = circuit.input(0, DIM, DIM);
    let c = circuit.multiply(a.clone(), b.clone());
    let d = circuit.output(c);
    circuit.add_endpoint(d.clone());

    a.lock().unwrap().set_input(&mat_a, 0);
    b.lock().unwrap().set_input(&mat_b, 0);

    circuit.read_offline_from_file(&mut party);
    circuit.run_online_with_benchmark(&mut party);
    circuit.print_stats(&party);

    // Correctness check
    let result = d.lock().unwrap().get_clear();
    let mut errors = 0;
    for (i, (&got, &exp)) in result.iter().zip(expected.iter()).enumerate() {
        if got != exp {
            if errors < 10 {
                let row = i / DIM;
                let col = i % DIM;
                eprintln!("MISMATCH at ({}, {}): got {}, expected {}", row, col, got, exp);
            }
            errors += 1;
        }
    }
    if errors == 0 {
        println!("Correctness check PASSED: all {} elements match", result.len());
    } else {
        println!("Correctness check FAILED: {}/{} elements wrong", errors, result.len());
    }
}
